IndexOutofBoundsException in argument-list-wrapping-rule formatting file with many corrections

## Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->

This test will throw an exception:

```
ArgumentListWrappingRule().format(
    """
        package com.foo
    
        class MyClass() {
            private fun doSomething() {
                if (d == 0 || e == 0f) {
                    c.ee(hh, d, d, yy)
                } else {
                    foo.blah()
                    val dr = t - u
                    val xx = -gg(dr) * rr(2f * d, dr.hh)
                    foo.bar(
                        dd,
                        g - d
                    )
                    foo.baz(
                        a.b, a.c - d
                    )
                    foo.biz(
                        a.b, a.c - d,
                        a.b, a.c,
                        a.b + d, a.c
                    )
                    foo.baz(
                        a.x - d, a.c
                    )
                    foo.biz(
                        a.x - d, a.c,
                        a.x, a.c,
                        a.x, a.c - d
                    )
                    foo.baz(
                        a.x, a.j + d
                    )
                }
            }
        }
    """.trimIndent()
```

```
Caused by: java.lang.IndexOutOfBoundsException: Wrong offset: 839. Should be in range: [0, 773]
	at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.LineSet.findLineIndex(LineSet.java:178)
	at org.jetbrains.kotlin.com.intellij.openapi.editor.impl.DocumentImpl.getLineNumber(DocumentImpl.java:991)
	at com.pinterest.ktlint.core.ast.PackageKt.lineNumber(package.kt:236)
	at com.pinterest.ktlint.ruleset.experimental.ArgumentListWrappingRule.isOnSameLineAsIfKeyword(ArgumentListWrappingRule.kt:225)
	at com.pinterest.ktlint.ruleset.experimental.ArgumentListWrappingRule.visit(ArgumentListWrappingRule.kt:96)
	at com.pinterest.ktlint.core.KtLint$format$1.invoke(KtLint.kt:300)
```

When you make it much shorter, it does not crash. Perhaps each node's startOffset is not being updated as corrections are made?

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* Version of ktlint used: 0.40.0
* Name and version (or code for custom task) of integration used (Gradle plugin, Maven plugin, command line, custom Gradle task): CLI
* Version of Gradle used (if applicable): 6.6
* Operating System and version: Mac OS 10.15.7

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - Fix false positives when declaration has tail comments (`spacing-between-declarations-with-comments`) ([#1053](https://github.com/pinterest/ktlint/issues/1053))
 - Fix false positive after `else` keyword (`argument-list-wrapping`) ([#1047](https://github.com/pinterest/ktlint/issues/1047))
 - Fix formatting with comments (`colon-spacing`) ([#1057](https://github.com/pinterest/ktlint/issues/1057))
+- Fix IndexOutOfBoundsException in `argument-list-wrapping-rule` formatting file with many corrections ([#1081](https://github.com/pinterest/ktlint/issues/1081))
 
 ### Changed
 - Update Gradle shadow plugin to `6.1.0` version

--- a/ktlint-ruleset-experimental/src/main/kotlin/com/pinterest/ktlint/ruleset/experimental/ArgumentListWrappingRule.kt
+++ b/ktlint-ruleset-experimental/src/main/kotlin/com/pinterest/ktlint/ruleset/experimental/ArgumentListWrappingRule.kt
@@ -14,7 +14,6 @@ import com.pinterest.ktlint.core.ast.isRoot
 import com.pinterest.ktlint.core.ast.isWhiteSpace
 import com.pinterest.ktlint.core.ast.isWhiteSpaceWithNewline
 import com.pinterest.ktlint.core.ast.lineIndent
-import com.pinterest.ktlint.core.ast.lineNumber
 import com.pinterest.ktlint.core.ast.prevLeaf
 import com.pinterest.ktlint.core.ast.visit
 import kotlin.math.max
@@ -234,10 +233,18 @@ class ArgumentListWrappingRule : Rule("argument-list-wrapping") {
     private fun ASTNode.isOnSameLineAsControlFlowKeyword(): Boolean {
         val containerNode = psi.getStrictParentOfType<KtContainerNode>() ?: return false
         if (containerNode.node.elementType == ELSE) return false
-        return when (val parent = containerNode.parent) {
-            is KtIfExpression, is KtWhileExpression -> parent.node
+        val controlFlowKeyword = when (val parent = containerNode.parent) {
+            is KtIfExpression -> parent.ifKeyword.node
+            is KtWhileExpression -> parent.firstChild.node
             is KtDoWhileExpression -> parent.whileKeyword?.node
             else -> null
-        }?.lineNumber() == lineNumber()
+        } ?: return false
+
+        var prevLeaf = prevLeaf() ?: return false
+        while (prevLeaf != controlFlowKeyword) {
+            if (prevLeaf.isWhiteSpaceWithNewline()) return false
+            prevLeaf = prevLeaf.prevLeaf() ?: return false
+        }
+        return true
     }
 }

--- a/ktlint-ruleset-experimental/src/test/kotlin/com/pinterest/ktlint/ruleset/experimental/ArgumentListWrappingRuleTest.kt
+++ b/ktlint-ruleset-experimental/src/test/kotlin/com/pinterest/ktlint/ruleset/experimental/ArgumentListWrappingRuleTest.kt
@@ -589,4 +589,99 @@ class ArgumentListWrappingRuleTest {
             )
         ).isEmpty()
     }
+
+    // https://github.com/pinterest/ktlint/issues/1081
+    @Test
+    fun testManyCorrections() {
+        assertThat(
+            ArgumentListWrappingRule().format(
+                """
+                package com.foo
+
+                class MyClass() {
+                    private fun doSomething() {
+                        if (d == 0 || e == 0f) {
+                            c.ee(hh, d, d, yy)
+                        } else {
+                            foo.blah()
+                            val dr = t - u
+                            val xx = -gg(dr) * rr(2f * d, dr.hh)
+                            foo.bar(
+                                dd,
+                                g - d
+                            )
+                            foo.baz(
+                                a.b, a.c - d
+                            )
+                            foo.biz(
+                                a.b, a.c - d,
+                                a.b, a.c,
+                                a.b + d, a.c
+                            )
+                            foo.baz(
+                                a.x - d, a.c
+                            )
+                            foo.biz(
+                                a.x - d, a.c,
+                                a.x, a.c,
+                                a.x, a.c - d
+                            )
+                            foo.baz(
+                                a.x, a.j + d
+                            )
+                        }
+                    }
+                }
+                """.trimIndent()
+            )
+        ).isEqualTo(
+            """
+            package com.foo
+
+            class MyClass() {
+                private fun doSomething() {
+                    if (d == 0 || e == 0f) {
+                        c.ee(hh, d, d, yy)
+                    } else {
+                        foo.blah()
+                        val dr = t - u
+                        val xx = -gg(dr) * rr(2f * d, dr.hh)
+                        foo.bar(
+                            dd,
+                            g - d
+                        )
+                        foo.baz(
+                            a.b,
+                            a.c - d
+                        )
+                        foo.biz(
+                            a.b,
+                            a.c - d,
+                            a.b,
+                            a.c,
+                            a.b + d,
+                            a.c
+                        )
+                        foo.baz(
+                            a.x - d,
+                            a.c
+                        )
+                        foo.biz(
+                            a.x - d,
+                            a.c,
+                            a.x,
+                            a.c,
+                            a.x,
+                            a.c - d
+                        )
+                        foo.baz(
+                            a.x,
+                            a.j + d
+                        )
+                    }
+                }
+            }
+            """.trimIndent()
+        )
+    }
 }

